xls/scheduling/pipeline_schedule.cc

```diff
@@ -141,6 +144,175 @@ absl::StatusOr<ScheduleCycleMap> ScheduleToMinimizeRegisters(
+    int64_t clock_period_ps) {
+  XLS_VLOG(3) << "ScheduleToMinimizeRegistersSDC()";
+  XLS_VLOG(3) << "  pipeline stages = " << pipeline_stages;
+  XLS_VLOG_LINES(4, f->DumpIr());
+
+  XLS_VLOG(4) << "Initial bounds:";
+  XLS_VLOG_LINES(4, bounds->ToString());
+
+  std::unique_ptr<MPSolver> solver(MPSolver::CreateSolver("SCIP"));
+  if (!solver) {
+    XLS_LOG(WARNING) << "SCIP solver unavailable.";
+    return absl::UnavailableError("SCIP solver unavailable.");
+  }
+
+  const double infinity = solver->infinity();
+
+  absl::flat_hash_map<Node*, MPVariable*>
+      node_cycle_var;  // Node's cycle after scheduling
+  absl::flat_hash_map<Node*, MPVariable*>
+      node_lifetime_var;  // Node's lifetime, since it finishes until it's
+                          // consumed by the last user
+  for (Node* node : f->nodes()) {
+    node_cycle_var[node] =
+        solver->MakeNumVar(bounds->lb(node), bounds->ub(node), node->GetName());
+    node_lifetime_var[node] = solver->MakeNumVar(
+        0.0, infinity, absl::StrFormat("lifetime_%s", node->GetName()));
+  }
+
+  for (Node* node : f->nodes()) {
+    MPVariable* const var_node = node_cycle_var.at(node);
+    MPVariable* const var_node_lifetime = node_lifetime_var.at(node);
+    for (Node* user : node->users()) {
+      MPVariable* const var_user = node_cycle_var.at(user);
+
+      // Constraint: cycle[i_user] - cycle[i] <= lifetime[i]
+      MPConstraint* const c = solver->MakeRowConstraint(-infinity, 0);
+      c->SetCoefficient(var_user, 1);
+      c->SetCoefficient(var_node, -1);
+      c->SetCoefficient(var_node_lifetime, -1);
+    }
+  }
+
+  XLS_ASSIGN_OR_RETURN(auto node_delay, ComputeNodeDelay(f, delay_estimator));
+  for (auto [path, path_delay] :
+       ComputeCriticalCombinationalPaths(f, node_delay)) {
+    if (path_delay < clock_period_ps) {
+      continue;
+    }
+    auto [src, dst] = path;
+    // Constraint: cycle[src] - cycle[dst] <= -path_delay/clock_period_ps
+    MPConstraint* const c =
+        solver->MakeRowConstraint(-infinity, -path_delay / clock_period_ps);
+    c->SetCoefficient(node_cycle_var[src], 1);
+    c->SetCoefficient(node_cycle_var[dst], -1);
+  }
+
+  MPObjective* const objective = solver->MutableObjective();
+  for (Node* node : f->nodes()) {
+    objective->SetCoefficient(node_lifetime_var[node],
+                              node->GetType()->GetFlatBitCount());
+  }
+  objective->SetMinimization();
+
+  const MPSolver::ResultStatus result_status = solver->Solve();
+  // Check that the problem has an optimal solution.
+  if (result_status != MPSolver::OPTIMAL) {
+    XLS_LOG(FATAL) << "The problem does not have an optimal solution!";
+    return absl::InternalError(
+        "The problem does not have an optimal solution!");
+  }
+
+  // extract result
+  ScheduleCycleMap cycle_map;
+  for (Node* node : f->nodes()) {
+    double cycle = node_cycle_var[node]->solution_value();
+    XLS_CHECK_LE(std::fabs(cycle - std::floor(cycle)), 1e-3)
```

Comment:
Shouldn't you round rather than floor here? Also, use `XLS_RET_CHECK_LE` or directly call absl::InternalError if that doesn't exist.